Fix issue 16774: talk about constructable stylesheets in the shadow D… (#29484)
mdn/content: files/en-us/web/api/web_components/using_shadow_dom/index.md

## Applying styles inside the shadow DOM

> To style page elements in the shadow DOM, we can:

Reviewer: ```suggestion
To style page elements in the shadow DOM, we can create a {{htmlelement("style")}} element and attach it to the shadow root, [include a `<style>`](/en-US/docs/Web/API/Web_components/Using_templates_and_slots#creating_a_template_with_some_slots) directly within a with {{htmlelement("template")}} , or  attach a style sheet to the shadow root. The rules within are scoped to the shadow DOM tree in all cases.

The steps to attach a style sheet to the shadow root include:
```

stating there is an alternative, then doing the step.

> - Create an empty {{domxref("CSSStyleSheet")}} object
> - Set its content using {{domxref("CSSStyleSheet.replace()")}} or {{domxref("CSSStyleSheet.replaceSync()")}}
> - Add it to the shadow root by assigning it to {{domxref("ShadowRoot.adoptedStyleSheets")}}

Reviewer: ```suggestion
1. Create an empty {{domxref("CSSStyleSheet")}} object
2. Set its content using {{domxref("CSSStyleSheet.replace()")}} or {{domxref("CSSStyleSheet.replaceSync()")}}
3. Add it to the shadow root by assigning it to {{domxref("ShadowRoot.adoptedStyleSheets")}}
```
These are steps, so numbering them

So the encapsulation of CSS works both ways: page styles don't affect the shadow DOM tree, and shadow DOM styles don't affect elements outside the shadow DOM.

> You can also apply styles inside a shadow DOM by creating a {{htmlelement("style")}} element and adding it to the shadow DOM tree. However, this is inefficient if you have many shadow DOM trees in a document, because the browser may need to parse the rules within every `<style>` element.

Reviewer: Not sure if this is accurate, but even if it is, I am not sure we should recommend against `<style>`, as it's the simplest way to create a component using `<template>`. Marinating on this.

> You can also apply styles inside a shadow DOM by creating a {{htmlelement("style")}} element and adding it to the shadow DOM tree. However, this is inefficient if you have many shadow DOM trees in a document, because the browser may need to parse the rules within every `<style>` element.
> 
> Creating a `CSSStyleSheet` and assigning it to the shadow root using `adoptedStyleSheets` allows us to create a single stylesheet and share it among many DOM trees.
> 
> For example, a component library could create a single stylesheet and then share it among all the custom elements belonging to that library.

Reviewer: ```suggestion
Applying CSS styles inside a shadow DOM by creating a {{htmlelement("style")}} element and adding it to the shadow DOM tree may seem inefficient if you have many shadow DOM trees in a document as the browser needs to parse each occurrence, browsers are extremely efficient at parsing CSS. There are likely lower-hanging fruit with a much greater impact on performance optimization available that are more impactful places to focus on to improve performance.

While adding a `<style>` block to a `<template>` means each occurrence of the custom element may include the same style block, the template is only downloaded once, meaning duplicate bytes are not an issue. And, again, browsers are optimized for parsing CSS.

Creating a `CSSStyleSheet` and assigning it to the shadow root using `adoptedStyleSheets` allows us to create a single stylesheet and share it among many DOM trees. Similarly, the `<style>` within the `<template>` is a single style block shared among each occurrence of the custom element.

For example, a component library could include styles directly within each custom element that are element-specific and create a single global stylesheet for the library that is shared among all the custom elements within that library.
```

- [Using custom elements](/en-US/docs/Web/API/Web_components/Using_custom_elements)
- [Using templates and slots](/en-US/docs/Web/API/Web_components/Using_templates_and_slots)
- {{domxref("Element.attachShadow()")}}
- {{domxref("ShadowRoot.adoptedStyleSheets")}}
- {{domxref("CSSStyleSheet.replace()")}}

> - {{domxref("CSSStyleSheet.replaceSync()")}}

Reviewer: ```suggestion
- {{domxref("CSSStyleSheet.replaceSync()")}}
- {{HTMLelement("template")}}
- [CSS scoping](/en-US/docs/Web/CSS/CSS_scoping) module
- {{CSSXref(":host")}}
- {{CSSXref(":host_function", ":host()")}}
- {{CSSXref(":host-context", ":host-context()")}}
- {{CSSXref("::slotted", "::slotted()")}}
- [CSS shadow parts](/en-US/docs/Web/CSS/CSS_shadow_parts) module
- {{CSSXref("::part")}}

```
